{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This NB was created to generate primers for KOing genes in yeast by replacing the gene with NanMX\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### the KanMX CDS is identical for all KOs, an upstream fragement and downstream fragment are created specific \n",
    "### for each gene\n",
    "\n",
    "### the upstram and dwostream fragement have approx. 400bp homologous to site to delete and 40bp homologous to the \n",
    "#KanMX cassette\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import  Bio\n",
    "from Bio.Seq import Seq\n",
    "import DNA_tools as dt\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import importlib #importlib.reload(ca)\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['vps34.gb',\n",
       " 'cox12.gb',\n",
       " 'shy1.gb',\n",
       " 'rcr2.gb',\n",
       " 'cax4.gb',\n",
       " 'nam7.gb',\n",
       " 'dld2.gb',\n",
       " 'yor1.gb',\n",
       " 'kanmx.gb',\n",
       " 'kgd2.gb',\n",
       " 'mip1.gb',\n",
       " 'yku70.gb',\n",
       " 'cur1.gb',\n",
       " 'sgm1.gb']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path2parts = '/Users/paucac/Documents/PhD/CAB/CRISPA/Betaxanthin/DNA_parts/KO_confirmations/'\n",
    "os.listdir(path2parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#some parameters:\n",
    "homology_length= 400 #length of the homology region upstream and downstream of a CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## dict storing file names of parts:\n",
    "genes2KO=['cax4', 'cox12','cur1','dld2','kgd2', 'mip1','nam7','rcr2','sgm1', 'shy1', 'vps34', 'yku70', 'yor1']\n",
    "genes2KO= [g.upper() for g in genes2KO]\n",
    "part_file={}\n",
    "for gene in genes2KO:\n",
    "    part_file[gene]= gene+\".gb\"\n",
    "\n",
    "part_seq={}\n",
    "part_seq['KanMX']='taggtctagagatctgtttagcttgcctcgtccccgccgggtcacccggccagcgacatggaggcccagaataccctccttgacagtcttgacgtgcgcagctcaggggcatgatgtgactgtcgcccgtacatttagcccatacatccccatgtataatcatttgcatccatacattttgatggccgcacggcgcgaagcaaaaattacggctcctcgctgcagacctgcgagcagggaaacgctcccctcacagacgcgttgaattgtccccacgccgcgcccctgtagagaaatataaaaggttaggatttgccactgaggttcttctttcatatacttccttttaaaatcttgctaggatacagttctcacatcacatccgaacataaacaaccatgggtaaggaaaagactcacgtttcgaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccggcaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataagcttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaatcagtactgacaataaaaagattcttgttttcaagaacttgtcatttgtatagtttttttatattgtagttgttctattttaatcaaatgttagcgtgatttatattttttttcgcctcgacatcatctgcccagatgcgaagttaagtgcgcagaaagtaatatcatgcgtcaatcgtatgtgaatgctggtcgctatactgctgtcgattcgatactaacgccgccatccagtgtcgaaaacgagctctcgagaacccttaat'\n",
    "\n",
    "#Defines the Overhangs with KanMX  \n",
    "KanMX_overhang_Up = dt.reverse_complement(part_seq['KanMX'][0:40]) #we copy the reverse complement for the upstream up\n",
    "KanMX_overhang_Dw = part_seq['KanMX'][-40:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract the sequence [upper homology:CDS:lower homology]\n",
    "\n",
    "for gene in genes2KO:\n",
    "    #open .gb file\n",
    "    start=0\n",
    "    stop=0\n",
    "    recs = [rec for rec in SeqIO.parse(path2parts + part_file[gene], \"genbank\")]\n",
    "    for rec in recs:\n",
    "        #loop through feature to features to find the location  of the CDS:\n",
    "        for feat in rec.features:\n",
    "            if gene + ' gene'in feat.qualifiers['label']:\n",
    "                location= feat.location\n",
    "                start = location.start.position\n",
    "                stop = location.end.position\n",
    "        #stores seq in dict\n",
    "        part_seq[gene] = str(rec.seq[start-homology_length:stop+homology_length])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#design the KanMX primers\n",
    "Kan_up= dt.make_anneal(part_seq['KanMX'].upper(), 'F' , 50)\n",
    "Kan_dw= dt.make_anneal(part_seq['KanMX'], 'R' , 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(dt)\n",
    "UF_list=[]\n",
    "UR_list=[]\n",
    "DF_list=[]\n",
    "DR_list=[]\n",
    "for gene in genes2KO:\n",
    "    #design the upstream F and R primers\n",
    "    seq= part_seq[gene]\n",
    "    UF = dt.make_anneal(seq, 'F' , 50)\n",
    "    UR = dt.make_anneal(seq[0:homology_length], 'R' , 50)\n",
    "    #design the downstream F and R primers:\n",
    "    DF = dt.make_anneal(seq[-homology_length:], 'F' , 50)\n",
    "    DR = dt.make_anneal(seq, 'R' , 50)\n",
    "\n",
    "    #copy the KanMX 30bp overhang to the UR and DF:\n",
    "    UR =  (( KanMX_overhang_Up).lower() + UR[0], UR[1])\n",
    "\n",
    "    DF =  ((KanMX_overhang_Dw).lower() + DF[0], DF[1])\n",
    "    \n",
    "    UF_list.append(UF)\n",
    "    UR_list.append(UR)\n",
    "    DF_list.append(DF)\n",
    "    DR_list.append(DR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort the ouptut into a nice data frame\n",
    "\n",
    "all_primers =[]\n",
    "all_temp =[]\n",
    "position = []\n",
    "Gene= []\n",
    "NA =[]\n",
    "Primer_name=[]\n",
    "\n",
    "for i, gene in enumerate(genes2KO):\n",
    "\n",
    "        \n",
    "    all_primers.append(UF_list[i][0])\n",
    "    all_temp.append(str(round(UF_list[i][1] ,1) +6 ) + ' phus')      #we add 6C for NEB phusion\n",
    "    position.append('UF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+1+3*i).zfill(3))\n",
    "\n",
    "    \n",
    "    all_primers.append(UR_list[i][0])\n",
    "    all_temp.append(str(round(UR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('UR')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+2+3*i).zfill(3))\n",
    "    \n",
    "    all_primers.append(DF_list[i][0])\n",
    "    all_temp.append(str(round(DF_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+3+3*i).zfill(3))\n",
    "    \n",
    "    all_primers.append(DR_list[i][0])\n",
    "    all_temp.append(str(round(DR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DR')               \n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+4+3*i).zfill(3))\n",
    " \n",
    "#add the KanMX primers\n",
    "all_primers.append(Kan_up[0])\n",
    "all_temp.append(str(round(Kan_up[1],1) + 6)+ ' phus')   \n",
    "position.append('F')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "all_primers.append(Kan_dw[0])\n",
    "all_temp.append(str(round(Kan_dw[1],1) + 6)+ ' phus')   \n",
    "position.append('R')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "\n",
    "#some primers are over IDT 60bp limit we shorten there homology region to KanMX:\n",
    "for i, primer in enumerate(all_primers):\n",
    "    if len(primer) >60:\n",
    "        all_primers[i] = primer[-60:]\n",
    "        if  (sum([bp.islower() for bp in all_primers[i]]) < 30):\n",
    "            print('WARNING your homology overhang was cut to <30bp')                    \n",
    "                            \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = np.stack([Primer_name, all_primers,position, all_temp,NA, Gene, NA, NA] ).T\n",
    "Primer_Table=pd.DataFrame(df , columns=['Primer Name', 'Seq','Dir', 'Tm (c)','User', 'Template', 'Template_Len','benchling'])\n",
    "\n",
    "Primer_Table.set_index('Primer Name')\n",
    "\n",
    "Primer_Table.to_csv('/Users/paucac/downloads/sheet2del.csv', sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Primer Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Dir</th>\n",
       "      <th>Tm (c)</th>\n",
       "      <th>User</th>\n",
       "      <th>Template</th>\n",
       "      <th>Template_Len</th>\n",
       "      <th>benchling</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BTX.001</td>\n",
       "      <td>ATTCTGCCGTAGGAAACTT</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BTX.002</td>\n",
       "      <td>cgaggcaagctaaacagatctctagacctaGTTGAACATTTCATTT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.8 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BTX.003</td>\n",
       "      <td>ccaagagctctcgagcaaaagctgtgacctaccgGATATGATCCGA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BTX.004</td>\n",
       "      <td>GCCAAACTAGTGGTACAGT</td>\n",
       "      <td>DR</td>\n",
       "      <td>57.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BTX.005</td>\n",
       "      <td>ATCTCAGTGAATTGAAAGTTCATGC</td>\n",
       "      <td>UF</td>\n",
       "      <td>58.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>COX12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BTX.006</td>\n",
       "      <td>gggacgaggcaagctaaacagatctctagacctaGGATCTTAGTTT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>57.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>COX12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BTX.007</td>\n",
       "      <td>caagagctctcgagcaaaagctgtgacctaccgGTCTATTATACTG...</td>\n",
       "      <td>DF</td>\n",
       "      <td>57.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>COX12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BTX.008</td>\n",
       "      <td>TTGGTGACAAGTGCTGC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>COX12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BTX.009</td>\n",
       "      <td>CAAGCACAAAATATGATAAAAAAGCA</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CUR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BTX.010</td>\n",
       "      <td>gcggggacgaggcaagctaaacagatctctagacctaTAGTCAGTT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CUR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BTX.011</td>\n",
       "      <td>taattcccaagagctctcgagcaaaagctgtgacctaccgGAAATA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.9 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CUR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BTX.012</td>\n",
       "      <td>ATTCCAACAATTTCTTTACCTGAT</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CUR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BTX.013</td>\n",
       "      <td>TCACCAGTAATGTCGCATAA</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.5 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>DLD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BTX.014</td>\n",
       "      <td>ggcggggacgaggcaagctaaacagatctctagacctaCTTGTTAT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>DLD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BTX.015</td>\n",
       "      <td>aagagctctcgagcaaaagctgtgacctaccgAAATAAAATGCCAA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>DLD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BTX.016</td>\n",
       "      <td>AATAATTTGCACTTACAACAAGAGC</td>\n",
       "      <td>DR</td>\n",
       "      <td>58.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>DLD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>BTX.017</td>\n",
       "      <td>AAAATTGTGTTTTTCAACATGTTAGG</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KGD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BTX.018</td>\n",
       "      <td>cggcggggacgaggcaagctaaacagatctctagacctaTTTGAAA...</td>\n",
       "      <td>UR</td>\n",
       "      <td>57.5 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KGD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BTX.019</td>\n",
       "      <td>aagagctctcgagcaaaagctgtgacctaccgTGTTTAAATAATTA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KGD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BTX.020</td>\n",
       "      <td>CCGTCAAAGACGATATTAACAA</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KGD2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BTX.021</td>\n",
       "      <td>TGGAGCAGCAGTATTTATACC</td>\n",
       "      <td>UF</td>\n",
       "      <td>58.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>MIP1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>BTX.022</td>\n",
       "      <td>ggcggggacgaggcaagctaaacagatctctagacctaGCAAAATC...</td>\n",
       "      <td>UR</td>\n",
       "      <td>58.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>MIP1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BTX.023</td>\n",
       "      <td>taattcccaagagctctcgagcaaaagctgtgacctaccgACTTGT...</td>\n",
       "      <td>DF</td>\n",
       "      <td>57.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>MIP1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BTX.024</td>\n",
       "      <td>CGTACTCTTGGTCAGCC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.6 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>MIP1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BTX.025</td>\n",
       "      <td>GTTTATTTTGTTGGCAGTTATATACA</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>NAM7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BTX.026</td>\n",
       "      <td>ccggcggggacgaggcaagctaaacagatctctagacctaTTCGGT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>NAM7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BTX.027</td>\n",
       "      <td>agagctctcgagcaaaagctgtgacctaccgTTGATAATGACAATG...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>NAM7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BTX.028</td>\n",
       "      <td>AATTCATGAACGGGAAATAAGAA</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>NAM7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BTX.029</td>\n",
       "      <td>ATATTTTTATCTACTGTGAGGAAAAGT</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.9 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RCR2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BTX.030</td>\n",
       "      <td>ggggacgaggcaagctaaacagatctctagacctaTTCAAAACTAT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.6 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RCR2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BTX.031</td>\n",
       "      <td>caagagctctcgagcaaaagctgtgacctaccgTTTTTCGTTTAAT...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RCR2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BTX.032</td>\n",
       "      <td>CATTTTCATTTTTTGAATTCAACGC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RCR2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BTX.033</td>\n",
       "      <td>GGATGAAGTTGACGATACTAGA</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SGM1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BTX.034</td>\n",
       "      <td>cgaggcaagctaaacagatctctagacctaCTAGAAATACACTCTT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SGM1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BTX.035</td>\n",
       "      <td>attcccaagagctctcgagcaaaagctgtgacctaccgAGCAAAAT...</td>\n",
       "      <td>DF</td>\n",
       "      <td>57.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SGM1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BTX.036</td>\n",
       "      <td>ACGAAGTCATGCTGTTAATGA</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SGM1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BTX.037</td>\n",
       "      <td>TGTTCGTCCCCCCA</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SHY1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BTX.038</td>\n",
       "      <td>cggggacgaggcaagctaaacagatctctagacctaTGCTTTCCTC...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SHY1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BTX.039</td>\n",
       "      <td>taattcccaagagctctcgagcaaaagctgtgacctaccgACGGCA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>58.8 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SHY1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BTX.040</td>\n",
       "      <td>CTTGTGGTCCCTAATTTAGC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.9 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>SHY1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BTX.041</td>\n",
       "      <td>TGTCTCCTGTATTTCCAATCC</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>VPS34</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BTX.042</td>\n",
       "      <td>ccggcggggacgaggcaagctaaacagatctctagacctaTGTTGG...</td>\n",
       "      <td>UR</td>\n",
       "      <td>58.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>VPS34</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BTX.043</td>\n",
       "      <td>aattcccaagagctctcgagcaaaagctgtgacctaccgTTGGTTG...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.5 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>VPS34</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BTX.044</td>\n",
       "      <td>CGGTGGGTTTTCTGAAATC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>VPS34</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>BTX.045</td>\n",
       "      <td>ATATAGGATAATTATCATCCCCTC</td>\n",
       "      <td>UF</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YKU70</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>BTX.046</td>\n",
       "      <td>gcggggacgaggcaagctaaacagatctctagacctaAATATTCCC...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YKU70</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>BTX.047</td>\n",
       "      <td>tcccaagagctctcgagcaaaagctgtgacctaccgAGACGATTGA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.7 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YKU70</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>BTX.048</td>\n",
       "      <td>GAAATTTCAAATTGGGCAAACAC</td>\n",
       "      <td>DR</td>\n",
       "      <td>57.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YKU70</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>BTX.049</td>\n",
       "      <td>CCGTGGAAATAGCCGG</td>\n",
       "      <td>UF</td>\n",
       "      <td>58.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YOR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>BTX.050</td>\n",
       "      <td>ccggcggggacgaggcaagctaaacagatctctagacctaATTCGT...</td>\n",
       "      <td>UR</td>\n",
       "      <td>57.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YOR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>BTX.051</td>\n",
       "      <td>agagctctcgagcaaaagctgtgacctaccgTTTATATTATTTGTT...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YOR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>BTX.052</td>\n",
       "      <td>TTCACTTCTTTTACGTATTTGAAATATC</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.8 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>YOR1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>BTX.XXX</td>\n",
       "      <td>TAGGTCTAGAGATCTGTTTAGC</td>\n",
       "      <td>F</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KanMX (pCfB 2312)</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>BTX.XXX</td>\n",
       "      <td>ATTAAGGGTTCTCGAGAGC</td>\n",
       "      <td>R</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KanMX (pCfB 2312)</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Primer Name                                                Seq Dir  \\\n",
       "0      BTX.001                                ATTCTGCCGTAGGAAACTT  UF   \n",
       "1      BTX.002  cgaggcaagctaaacagatctctagacctaGTTGAACATTTCATTT...  UR   \n",
       "2      BTX.003  ccaagagctctcgagcaaaagctgtgacctaccgGATATGATCCGA...  DF   \n",
       "3      BTX.004                                GCCAAACTAGTGGTACAGT  DR   \n",
       "4      BTX.005                          ATCTCAGTGAATTGAAAGTTCATGC  UF   \n",
       "5      BTX.006  gggacgaggcaagctaaacagatctctagacctaGGATCTTAGTTT...  UR   \n",
       "6      BTX.007  caagagctctcgagcaaaagctgtgacctaccgGTCTATTATACTG...  DF   \n",
       "7      BTX.008                                  TTGGTGACAAGTGCTGC  DR   \n",
       "8      BTX.009                         CAAGCACAAAATATGATAAAAAAGCA  UF   \n",
       "9      BTX.010  gcggggacgaggcaagctaaacagatctctagacctaTAGTCAGTT...  UR   \n",
       "10     BTX.011  taattcccaagagctctcgagcaaaagctgtgacctaccgGAAATA...  DF   \n",
       "11     BTX.012                           ATTCCAACAATTTCTTTACCTGAT  DR   \n",
       "12     BTX.013                               TCACCAGTAATGTCGCATAA  UF   \n",
       "13     BTX.014  ggcggggacgaggcaagctaaacagatctctagacctaCTTGTTAT...  UR   \n",
       "14     BTX.015  aagagctctcgagcaaaagctgtgacctaccgAAATAAAATGCCAA...  DF   \n",
       "15     BTX.016                          AATAATTTGCACTTACAACAAGAGC  DR   \n",
       "16     BTX.017                         AAAATTGTGTTTTTCAACATGTTAGG  UF   \n",
       "17     BTX.018  cggcggggacgaggcaagctaaacagatctctagacctaTTTGAAA...  UR   \n",
       "18     BTX.019  aagagctctcgagcaaaagctgtgacctaccgTGTTTAAATAATTA...  DF   \n",
       "19     BTX.020                             CCGTCAAAGACGATATTAACAA  DR   \n",
       "20     BTX.021                              TGGAGCAGCAGTATTTATACC  UF   \n",
       "21     BTX.022  ggcggggacgaggcaagctaaacagatctctagacctaGCAAAATC...  UR   \n",
       "22     BTX.023  taattcccaagagctctcgagcaaaagctgtgacctaccgACTTGT...  DF   \n",
       "23     BTX.024                                  CGTACTCTTGGTCAGCC  DR   \n",
       "24     BTX.025                         GTTTATTTTGTTGGCAGTTATATACA  UF   \n",
       "25     BTX.026  ccggcggggacgaggcaagctaaacagatctctagacctaTTCGGT...  UR   \n",
       "26     BTX.027  agagctctcgagcaaaagctgtgacctaccgTTGATAATGACAATG...  DF   \n",
       "27     BTX.028                            AATTCATGAACGGGAAATAAGAA  DR   \n",
       "28     BTX.029                        ATATTTTTATCTACTGTGAGGAAAAGT  UF   \n",
       "29     BTX.030  ggggacgaggcaagctaaacagatctctagacctaTTCAAAACTAT...  UR   \n",
       "30     BTX.031  caagagctctcgagcaaaagctgtgacctaccgTTTTTCGTTTAAT...  DF   \n",
       "31     BTX.032                          CATTTTCATTTTTTGAATTCAACGC  DR   \n",
       "32     BTX.033                             GGATGAAGTTGACGATACTAGA  UF   \n",
       "33     BTX.034  cgaggcaagctaaacagatctctagacctaCTAGAAATACACTCTT...  UR   \n",
       "34     BTX.035  attcccaagagctctcgagcaaaagctgtgacctaccgAGCAAAAT...  DF   \n",
       "35     BTX.036                              ACGAAGTCATGCTGTTAATGA  DR   \n",
       "36     BTX.037                                     TGTTCGTCCCCCCA  UF   \n",
       "37     BTX.038  cggggacgaggcaagctaaacagatctctagacctaTGCTTTCCTC...  UR   \n",
       "38     BTX.039  taattcccaagagctctcgagcaaaagctgtgacctaccgACGGCA...  DF   \n",
       "39     BTX.040                               CTTGTGGTCCCTAATTTAGC  DR   \n",
       "40     BTX.041                              TGTCTCCTGTATTTCCAATCC  UF   \n",
       "41     BTX.042  ccggcggggacgaggcaagctaaacagatctctagacctaTGTTGG...  UR   \n",
       "42     BTX.043  aattcccaagagctctcgagcaaaagctgtgacctaccgTTGGTTG...  DF   \n",
       "43     BTX.044                                CGGTGGGTTTTCTGAAATC  DR   \n",
       "44     BTX.045                           ATATAGGATAATTATCATCCCCTC  UF   \n",
       "45     BTX.046  gcggggacgaggcaagctaaacagatctctagacctaAATATTCCC...  UR   \n",
       "46     BTX.047  tcccaagagctctcgagcaaaagctgtgacctaccgAGACGATTGA...  DF   \n",
       "47     BTX.048                            GAAATTTCAAATTGGGCAAACAC  DR   \n",
       "48     BTX.049                                   CCGTGGAAATAGCCGG  UF   \n",
       "49     BTX.050  ccggcggggacgaggcaagctaaacagatctctagacctaATTCGT...  UR   \n",
       "50     BTX.051  agagctctcgagcaaaagctgtgacctaccgTTTATATTATTTGTT...  DF   \n",
       "51     BTX.052                       TTCACTTCTTTTACGTATTTGAAATATC  DR   \n",
       "52     BTX.XXX                             TAGGTCTAGAGATCTGTTTAGC   F   \n",
       "53     BTX.XXX                                ATTAAGGGTTCTCGAGAGC   R   \n",
       "\n",
       "       Tm (c) User           Template Template_Len benchling  \n",
       "0   56.1 phus   na               CAX4           na        na  \n",
       "1   56.8 phus   na               CAX4           na        na  \n",
       "2   56.4 phus   na               CAX4           na        na  \n",
       "3   57.2 phus   na               CAX4           na        na  \n",
       "4   58.4 phus   na              COX12           na        na  \n",
       "5   57.0 phus   na              COX12           na        na  \n",
       "6   57.0 phus   na              COX12           na        na  \n",
       "7   56.2 phus   na              COX12           na        na  \n",
       "8   57.2 phus   na               CUR1           na        na  \n",
       "9   56.1 phus   na               CUR1           na        na  \n",
       "10  56.9 phus   na               CUR1           na        na  \n",
       "11  56.4 phus   na               CUR1           na        na  \n",
       "12  56.5 phus   na               DLD2           na        na  \n",
       "13  56.3 phus   na               DLD2           na        na  \n",
       "14  56.3 phus   na               DLD2           na        na  \n",
       "15  58.1 phus   na               DLD2           na        na  \n",
       "16  57.3 phus   na               KGD2           na        na  \n",
       "17  57.5 phus   na               KGD2           na        na  \n",
       "18  56.7 phus   na               KGD2           na        na  \n",
       "19  56.4 phus   na               KGD2           na        na  \n",
       "20  58.2 phus   na               MIP1           na        na  \n",
       "21  58.0 phus   na               MIP1           na        na  \n",
       "22  57.1 phus   na               MIP1           na        na  \n",
       "23  56.6 phus   na               MIP1           na        na  \n",
       "24  57.3 phus   na               NAM7           na        na  \n",
       "25  56.1 phus   na               NAM7           na        na  \n",
       "26  56.1 phus   na               NAM7           na        na  \n",
       "27  56.2 phus   na               NAM7           na        na  \n",
       "28  56.9 phus   na               RCR2           na        na  \n",
       "29  56.6 phus   na               RCR2           na        na  \n",
       "30  56.2 phus   na               RCR2           na        na  \n",
       "31  56.1 phus   na               RCR2           na        na  \n",
       "32  56.7 phus   na               SGM1           na        na  \n",
       "33  56.7 phus   na               SGM1           na        na  \n",
       "34  57.0 phus   na               SGM1           na        na  \n",
       "35  56.3 phus   na               SGM1           na        na  \n",
       "36  56.7 phus   na               SHY1           na        na  \n",
       "37  56.4 phus   na               SHY1           na        na  \n",
       "38  58.8 phus   na               SHY1           na        na  \n",
       "39  56.9 phus   na               SHY1           na        na  \n",
       "40  57.4 phus   na              VPS34           na        na  \n",
       "41  58.3 phus   na              VPS34           na        na  \n",
       "42  56.5 phus   na              VPS34           na        na  \n",
       "43  56.4 phus   na              VPS34           na        na  \n",
       "44  56.7 phus   na              YKU70           na        na  \n",
       "45  56.0 phus   na              YKU70           na        na  \n",
       "46  56.7 phus   na              YKU70           na        na  \n",
       "47  57.4 phus   na              YKU70           na        na  \n",
       "48  58.4 phus   na               YOR1           na        na  \n",
       "49  57.1 phus   na               YOR1           na        na  \n",
       "50  56.3 phus   na               YOR1           na        na  \n",
       "51  56.8 phus   na               YOR1           na        na  \n",
       "52  56.3 phus   na  KanMX (pCfB 2312)           na        na  \n",
       "53  56.1 phus   na  KanMX (pCfB 2312)           na        na  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Primer_Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a PCR table\n",
    "tube_label = []\n",
    "template = []\n",
    "tm=[]\n",
    "Fwd = []\n",
    "Rev = []\n",
    "Len = []\n",
    "for i, gene in enumerate(genes2KO):\n",
    "    \n",
    "    #upstream PCR\n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.up')\n",
    "    Len.append(homology_length)\n",
    "    #downstream PCR\n",
    "    \n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.dw')\n",
    "    Len.append(homology_length)\n",
    "    \n",
    "df = np.stack([template, tube_label, tm , Len, Fwd, Rev ] ).T\n",
    "PCR_Table=pd.DataFrame(df , columns=[ 'Template', 'tube_label', 'tm','Product (bp)', 'Fwd','Rev'])\n",
    "\n",
    "PCR_Table.set_index('Template')\n",
    "\n",
    "PCR_Table.to_csv('/Users/paucac/downloads/sheet2del2.csv', sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create primers for KOing TAL1 and ZWF1 in CCM strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['zwf1.gb', 'tal1.gb']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "path2parts = '/Users/paucac/Documents/PhD/CAB/CRISPA/CCM/DNA_parts/KO_confirmations/'\n",
    "os.listdir(path2parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## dict storing file names of parts:\n",
    "genes2KO=['zwf1','tal1']\n",
    "genes2KO= [g.upper() for g in genes2KO]\n",
    "part_file={}\n",
    "for gene in genes2KO:\n",
    "    part_file[gene]= gene+\".gb\"\n",
    "\n",
    "part_seq={}\n",
    "part_seq['KanMX']='taggtctagagatctgtttagcttgcctcgtccccgccgggtcacccggccagcgacatggaggcccagaataccctccttgacagtcttgacgtgcgcagctcaggggcatgatgtgactgtcgcccgtacatttagcccatacatccccatgtataatcatttgcatccatacattttgatggccgcacggcgcgaagcaaaaattacggctcctcgctgcagacctgcgagcagggaaacgctcccctcacagacgcgttgaattgtccccacgccgcgcccctgtagagaaatataaaaggttaggatttgccactgaggttcttctttcatatacttccttttaaaatcttgctaggatacagttctcacatcacatccgaacataaacaaccatgggtaaggaaaagactcacgtttcgaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccggcaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataagcttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaatcagtactgacaataaaaagattcttgttttcaagaacttgtcatttgtatagtttttttatattgtagttgttctattttaatcaaatgttagcgtgatttatattttttttcgcctcgacatcatctgcccagatgcgaagttaagtgcgcagaaagtaatatcatgcgtcaatcgtatgtgaatgctggtcgctatactgctgtcgattcgatactaacgccgccatccagtgtcgaaaacgagctctcgagaacccttaat'\n",
    "\n",
    "#Defines the Overhangs with KanMX  \n",
    "KanMX_overhang_Up = dt.reverse_complement(part_seq['KanMX'][0:40]) #we copy the reverse complement for the upstream up\n",
    "KanMX_overhang_Dw = part_seq['KanMX'][-40:]\n",
    "\n",
    "\n",
    "#extract the sequence [upper homology:CDS:lower homology]\n",
    "\n",
    "for gene in genes2KO:\n",
    "    #open .gb file\n",
    "    start=0\n",
    "    stop=0\n",
    "    recs = [rec for rec in SeqIO.parse(path2parts + part_file[gene], \"genbank\")]\n",
    "    for rec in recs:\n",
    "        #loop through feature to features to find the location  of the CDS:\n",
    "        for feat in rec.features:\n",
    "            if gene + ' gene'in feat.qualifiers['label']:\n",
    "                location= feat.location\n",
    "                start = location.start.position\n",
    "                stop = location.end.position\n",
    "        #stores seq in dict\n",
    "        part_seq[gene] = str(rec.seq[start-homology_length:stop+homology_length])\n",
    "\n",
    "\n",
    "        \n",
    "#design the KanMX primers\n",
    "Kan_up= dt.make_anneal(part_seq['KanMX'].upper(), 'F' , 50)\n",
    "Kan_dw= dt.make_anneal(part_seq['KanMX'], 'R' , 50)\n",
    "\n",
    "#designs the primers for KO:\n",
    "importlib.reload(dt)\n",
    "UF_list=[]\n",
    "UR_list=[]\n",
    "DF_list=[]\n",
    "DR_list=[]\n",
    "for gene in genes2KO:\n",
    "    #design the upstream F and R primers\n",
    "    seq= part_seq[gene]\n",
    "    UF = dt.make_anneal(seq, 'F' , 50)\n",
    "    UR = dt.make_anneal(seq[0:homology_length], 'R' , 50)\n",
    "    #design the downstream F and R primers:\n",
    "    DF = dt.make_anneal(seq[-homology_length:], 'F' , 50)\n",
    "    DR = dt.make_anneal(seq, 'R' , 50)\n",
    "\n",
    "    #copy the KanMX 30bp overhang to the UR and DF:\n",
    "    UR =  (( KanMX_overhang_Up).lower() + UR[0], UR[1])\n",
    "\n",
    "    DF =  ((KanMX_overhang_Dw).lower() + DF[0], DF[1])\n",
    "    \n",
    "    UF_list.append(UF)\n",
    "    UR_list.append(UR)\n",
    "    DF_list.append(DF)\n",
    "    DR_list.append(DR)\n",
    "    \n",
    "#sort the ouptut into a nice data frame\n",
    "\n",
    "all_primers =[]\n",
    "all_temp =[]\n",
    "position = []\n",
    "Gene= []\n",
    "NA =[]\n",
    "Primer_name=[]\n",
    "\n",
    "primer_n = 77 #to match your current primer indexing\n",
    "for i, gene in enumerate(genes2KO):\n",
    "\n",
    "        \n",
    "    all_primers.append(UF_list[i][0])\n",
    "    all_temp.append(str(round(UF_list[i][1] ,1) +6 ) + ' phus')      #we add 6C for NEB phusion\n",
    "    position.append('UF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+1+3*i +primer_n ).zfill(3))\n",
    "\n",
    "    \n",
    "    all_primers.append(UR_list[i][0])\n",
    "    all_temp.append(str(round(UR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('UR')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+2+3*i +primer_n).zfill(3))\n",
    "    \n",
    "    all_primers.append(DF_list[i][0])\n",
    "    all_temp.append(str(round(DF_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+3+3*i +primer_n).zfill(3))\n",
    "    \n",
    "    all_primers.append(DR_list[i][0])\n",
    "    all_temp.append(str(round(DR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DR')               \n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+4+3*i +primer_n).zfill(3))\n",
    " \n",
    "#add the KanMX primers\n",
    "all_primers.append(Kan_up[0])\n",
    "all_temp.append(str(round(Kan_up[1],1) + 6)+ ' phus')   \n",
    "position.append('F')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "all_primers.append(Kan_dw[0])\n",
    "all_temp.append(str(round(Kan_dw[1],1) + 6)+ ' phus')   \n",
    "position.append('R')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "\n",
    "#some primers are over IDT 60bp limit we shorten there homology region to KanMX:\n",
    "for i, primer in enumerate(all_primers):\n",
    "    if len(primer) >60:\n",
    "        all_primers[i] = primer[-60:]\n",
    "        if  (sum([bp.islower() for bp in all_primers[i]]) < 30):\n",
    "            print('WARNING your homology overhang was cut to <30bp')                    \n",
    "                            \n",
    "df = np.stack([Primer_name, all_primers,position, all_temp,NA, Gene, NA, NA] ).T\n",
    "Primer_Table=pd.DataFrame(df , columns=['Primer Name', 'Seq','Dir', 'Tm (c)','User', 'Template', 'Template_Len','benchling'])\n",
    "\n",
    "Primer_Table.set_index('Primer Name')\n",
    "\n",
    "Primer_Table.to_csv('/Users/paucac/downloads/sheet2del.csv', sep=',')        \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DNAEnv",
   "language": "python",
   "name": "dnaenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
