{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This NB was created to generate primers for KOing genes in yeast by replacing the gene with NanMX\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### the KanMX CDS is identical for all KOs, an upstream fragement and downstream fragment are created specific \n",
    "### for each gene\n",
    "\n",
    "### the upstram and dwostream fragement have approx. 400bp homologous to site to delete and 40bp homologous to the \n",
    "#KanMX cassette\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import  Bio\n",
    "from Bio.Seq import Seq\n",
    "import DNA_tools as dt\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import importlib #importlib.reload(ca)\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['vps34.gb',\n",
       " 'cox12.gb',\n",
       " 'shy1.gb',\n",
       " 'rcr2.gb',\n",
       " 'cax4.gb',\n",
       " 'nam7.gb',\n",
       " 'dld2.gb',\n",
       " 'yor1.gb',\n",
       " 'kanmx.gb',\n",
       " 'kgd2.gb',\n",
       " 'mip1.gb',\n",
       " 'yer084w.gb',\n",
       " 'yku70.gb',\n",
       " 'cur1.gb',\n",
       " 'rim13.gb',\n",
       " 'sgm1.gb']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path2parts = '/Users/paucac/Documents/PhD/CAB/CRISPA/Betaxanthin/DNA_parts/KO_confirmations/'\n",
    "os.listdir(path2parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#some parameters:\n",
    "homology_length= 400 #length of the homology region upstream and downstream of a CDS\n",
    "out_gap= 170 #length outside of integration sequences used as gap for confirmation primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "## dict storing file names of parts:\n",
    "genes2KO=['cax4', 'cox12','cur1','dld2','kgd2', 'mip1','nam7','rcr2','sgm1', 'shy1', 'vps34', 'yku70', 'yor1','yer084w','rim13']\n",
    "genes2KO= [g.upper() for g in genes2KO]\n",
    "part_file={}\n",
    "for gene in genes2KO:\n",
    "    part_file[gene]= gene+\".gb\"\n",
    "\n",
    "part_seq={}\n",
    "part_seq['KanMX']='taggtctagagatctgtttagcttgcctcgtccccgccgggtcacccggccagcgacatggaggcccagaataccctccttgacagtcttgacgtgcgcagctcaggggcatgatgtgactgtcgcccgtacatttagcccatacatccccatgtataatcatttgcatccatacattttgatggccgcacggcgcgaagcaaaaattacggctcctcgctgcagacctgcgagcagggaaacgctcccctcacagacgcgttgaattgtccccacgccgcgcccctgtagagaaatataaaaggttaggatttgccactgaggttcttctttcatatacttccttttaaaatcttgctaggatacagttctcacatcacatccgaacataaacaaccatgggtaaggaaaagactcacgtttcgaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccggcaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataagcttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaatcagtactgacaataaaaagattcttgttttcaagaacttgtcatttgtatagtttttttatattgtagttgttctattttaatcaaatgttagcgtgatttatattttttttcgcctcgacatcatctgcccagatgcgaagttaagtgcgcagaaagtaatatcatgcgtcaatcgtatgtgaatgctggtcgctatactgctgtcgattcgatactaacgccgccatccagtgtcgaaaacgagctctcgagaacccttaat'\n",
    "\n",
    "#Defines the Overhangs with KanMX  \n",
    "KanMX_overhang_Up = dt.reverse_complement(part_seq['KanMX'][0:40]) #we copy the reverse complement for the upstream up\n",
    "KanMX_overhang_Dw = part_seq['KanMX'][-40:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-570"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "-homology_length-out_gap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAX4 308 2168\n",
      "COX12 106 1498\n",
      "CUR1 30 1929\n",
      "DLD2 193 2926\n",
      "KGD2 9 2541\n",
      "MIP1 88 4993\n",
      "NAM7 179 4235\n",
      "RCR2 214 1987\n",
      "SGM1 19 3283\n",
      "SHY1 285 2595\n",
      "VPS34 36 3804\n",
      "YKU70 60 3009\n",
      "YOR1 528 6102\n",
      "YER084W 246 1773\n",
      "RIM13 93 3417\n"
     ]
    }
   ],
   "source": [
    "#extract the sequence [upper homology:CDS:lower homology]\n",
    "\n",
    "for gene in genes2KO:\n",
    "\n",
    "    #open .gb file\n",
    "    start=0\n",
    "    stop=0\n",
    "    recs = [rec for rec in SeqIO.parse(path2parts + part_file[gene], \"genbank\")]\n",
    "    for rec in recs:\n",
    "        #loop through feature to features to find the location  of the CDS:\n",
    "        for feat in rec.features:\n",
    "            if gene + ' gene'in feat.qualifiers['label'] or gene in feat.qualifiers['label']:\n",
    "                location= feat.location\n",
    "                start = location.start.position\n",
    "                stop = location.end.position\n",
    "\n",
    "        #stores seq in dict\n",
    "        part_seq[gene] = str(rec.seq[start-homology_length-out_gap:stop+homology_length+out_gap])\n",
    "        print(gene, start-homology_length-out_gap,stop+homology_length+out_gap)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "#design the KanMX primers\n",
    "Kan_up= dt.make_anneal(part_seq['KanMX'].upper(), 'F' , 50)\n",
    "Kan_dw= dt.make_anneal(part_seq['KanMX'], 'R' , 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAX4\n",
      "COX12\n",
      "CUR1\n",
      "DLD2\n",
      "KGD2\n",
      "MIP1\n",
      "NAM7\n",
      "RCR2\n",
      "SGM1\n",
      "SHY1\n",
      "VPS34\n",
      "YKU70\n",
      "YOR1\n",
      "YER084W\n",
      "RIM13\n"
     ]
    }
   ],
   "source": [
    "importlib.reload(dt)\n",
    "UF_list=[]\n",
    "UR_list=[]\n",
    "DF_list=[]\n",
    "DR_list=[]\n",
    "for gene in genes2KO:\n",
    "    #design the upstream F and R primers\n",
    "    seq= part_seq[gene]\n",
    "    UF = dt.make_anneal(seq, 'F' , 50)\n",
    "    UR = dt.make_anneal(seq[0:homology_length], 'R' , 50)\n",
    "    #design the downstream F and R primers:\n",
    "    DF = dt.make_anneal(seq[-homology_length:], 'F' , 50)\n",
    "    DR = dt.make_anneal(seq, 'R' , 50)\n",
    "\n",
    "    #copy the KanMX 30bp overhang to the UR and DF:\n",
    "    UR =  (( KanMX_overhang_Up).lower() + UR[0], UR[1])\n",
    "\n",
    "    DF =  ((KanMX_overhang_Dw).lower() + DF[0], DF[1])\n",
    "    \n",
    "    UF_list.append(UF)\n",
    "    UR_list.append(UR)\n",
    "    DF_list.append(DF)\n",
    "    DR_list.append(DR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(dt)\n",
    "UF_list=[]\n",
    "UR_list=[]\n",
    "DF_list=[]\n",
    "DR_list=[]\n",
    "for gene in genes2KO:\n",
    "    #design the upstream F and R primers\n",
    "    seq= part_seq[gene]\n",
    "    UF = dt.make_anneal(seq, 'F' , 50)\n",
    "    #design primers from leaving an upstream outgap and homology_length bp to CDS\n",
    "    UR = dt.make_anneal(seq[out_gap:homology_length], 'R' , 50)\n",
    "    #design the downstream F and R primers:\n",
    "    DF = dt.make_anneal(seq[-homology_length-out_gap:], 'F' , 50)\n",
    "    DR = dt.make_anneal(seq, 'R' , 50)\n",
    "\n",
    "    #copy the KanMX 30bp overhang to the UR and DF:\n",
    "    UR =  (( KanMX_overhang_Up).lower() + UR[0], UR[1])\n",
    "\n",
    "    DF =  ((KanMX_overhang_Dw).lower() + DF[0], DF[1])\n",
    "    \n",
    "    UF_list.append(UF)\n",
    "    UR_list.append(UR)\n",
    "    DF_list.append(DF)\n",
    "    DR_list.append(DR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING your homology overhang was cut to <30bp\n"
     ]
    }
   ],
   "source": [
    "#sort the ouptut into a nice data frame\n",
    "\n",
    "all_primers =[]\n",
    "all_temp =[]\n",
    "position = []\n",
    "Gene= []\n",
    "NA =[]\n",
    "Primer_name=[]\n",
    "primer_n = 0\n",
    "\n",
    "for i, gene in enumerate(genes2KO):\n",
    "\n",
    "        \n",
    "    all_primers.append(UF_list[i][0])\n",
    "    all_temp.append(str(round(UF_list[i][1] ,1) +6 ) + ' phus')      #we add 6C for NEB phusion\n",
    "    position.append('UF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+1+3*i +primer_n ).zfill(3))\n",
    "\n",
    "    \n",
    "    all_primers.append(UR_list[i][0])\n",
    "    all_temp.append(str(round(UR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('UR')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+2+3*i +primer_n ).zfill(3))\n",
    "    \n",
    "    all_primers.append(DF_list[i][0])\n",
    "    all_temp.append(str(round(DF_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+3+3*i +primer_n ).zfill(3))\n",
    "    \n",
    "    all_primers.append(DR_list[i][0])\n",
    "    all_temp.append(str(round(DR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DR')               \n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('BTX.'+str(i+4+3*i +primer_n ).zfill(3))\n",
    " \n",
    "#add the KanMX primers\n",
    "all_primers.append(Kan_up[0])\n",
    "all_temp.append(str(round(Kan_up[1],1) + 6)+ ' phus')   \n",
    "position.append('F')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "all_primers.append(Kan_dw[0])\n",
    "all_temp.append(str(round(Kan_dw[1],1) + 6)+ ' phus')   \n",
    "position.append('R')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "\n",
    "#some primers are over IDT 60bp limit we shorten there homology region to KanMX:\n",
    "for i, primer in enumerate(all_primers):\n",
    "    if len(primer) >60:\n",
    "        all_primers[i] = primer[-60:]\n",
    "        if  (sum([bp.islower() for bp in all_primers[i]]) < 30):\n",
    "            print('WARNING your homology overhang was cut to <30bp')                    \n",
    "                            \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = np.stack([Primer_name, all_primers,position, all_temp,NA, Gene, NA, NA] ).T\n",
    "Primer_Table=pd.DataFrame(df , columns=['Primer Name', 'Seq','Dir', 'Tm (c)','User', 'Template', 'Template_Len','benchling'])\n",
    "\n",
    "Primer_Table.set_index('Primer Name')\n",
    "\n",
    "Primer_Table.to_csv('/Users/paucac/downloads/sheet2del.csv', sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Primer Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Dir</th>\n",
       "      <th>Tm (c)</th>\n",
       "      <th>User</th>\n",
       "      <th>Template</th>\n",
       "      <th>Template_Len</th>\n",
       "      <th>benchling</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BTX.001</td>\n",
       "      <td>ATATATTTTATATATTGCCCTCAAAACT</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.0 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BTX.002</td>\n",
       "      <td>ccggcggggacgaggcaagctaaacagatctctagacctaACTGTG...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.2 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BTX.003</td>\n",
       "      <td>cagtgtcgaaaacgagctctcgagaacccttaatGATATGATCCGA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.4 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BTX.004</td>\n",
       "      <td>AAACACGCTCTTGAAATTTTTAAAG</td>\n",
       "      <td>DR</td>\n",
       "      <td>56.6 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>CAX4</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BTX.005</td>\n",
       "      <td>TCAGGCTGTTGATGAAGGT</td>\n",
       "      <td>UF</td>\n",
       "      <td>57.6 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>COX12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>BTX.058</td>\n",
       "      <td>ggacgaggcaagctaaacagatctctagacctaAATCTATTTAACC...</td>\n",
       "      <td>UR</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RIM13</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>BTX.059</td>\n",
       "      <td>catccagtgtcgaaaacgagctctcgagaacccttaatTTTTGATA...</td>\n",
       "      <td>DF</td>\n",
       "      <td>56.6 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RIM13</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>BTX.060</td>\n",
       "      <td>AGCTTCCTATTGTTAAAGAGAAGG</td>\n",
       "      <td>DR</td>\n",
       "      <td>58.5 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>RIM13</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>BTX.XXX</td>\n",
       "      <td>TAGGTCTAGAGATCTGTTTAGC</td>\n",
       "      <td>F</td>\n",
       "      <td>56.3 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KanMX (pCfB 2312)</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>BTX.XXX</td>\n",
       "      <td>ATTAAGGGTTCTCGAGAGC</td>\n",
       "      <td>R</td>\n",
       "      <td>56.1 phus</td>\n",
       "      <td>na</td>\n",
       "      <td>KanMX (pCfB 2312)</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Primer Name                                                Seq Dir  \\\n",
       "0      BTX.001                       ATATATTTTATATATTGCCCTCAAAACT  UF   \n",
       "1      BTX.002  ccggcggggacgaggcaagctaaacagatctctagacctaACTGTG...  UR   \n",
       "2      BTX.003  cagtgtcgaaaacgagctctcgagaacccttaatGATATGATCCGA...  DF   \n",
       "3      BTX.004                          AAACACGCTCTTGAAATTTTTAAAG  DR   \n",
       "4      BTX.005                                TCAGGCTGTTGATGAAGGT  UF   \n",
       "..         ...                                                ...  ..   \n",
       "57     BTX.058  ggacgaggcaagctaaacagatctctagacctaAATCTATTTAACC...  UR   \n",
       "58     BTX.059  catccagtgtcgaaaacgagctctcgagaacccttaatTTTTGATA...  DF   \n",
       "59     BTX.060                           AGCTTCCTATTGTTAAAGAGAAGG  DR   \n",
       "60     BTX.XXX                             TAGGTCTAGAGATCTGTTTAGC   F   \n",
       "61     BTX.XXX                                ATTAAGGGTTCTCGAGAGC   R   \n",
       "\n",
       "       Tm (c) User           Template Template_Len benchling  \n",
       "0   57.0 phus   na               CAX4           na        na  \n",
       "1   56.2 phus   na               CAX4           na        na  \n",
       "2   56.4 phus   na               CAX4           na        na  \n",
       "3   56.6 phus   na               CAX4           na        na  \n",
       "4   57.6 phus   na              COX12           na        na  \n",
       "..        ...  ...                ...          ...       ...  \n",
       "57  56.3 phus   na              RIM13           na        na  \n",
       "58  56.6 phus   na              RIM13           na        na  \n",
       "59  58.5 phus   na              RIM13           na        na  \n",
       "60  56.3 phus   na  KanMX (pCfB 2312)           na        na  \n",
       "61  56.1 phus   na  KanMX (pCfB 2312)           na        na  \n",
       "\n",
       "[62 rows x 8 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Primer_Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a PCR table\n",
    "tube_label = []\n",
    "template = []\n",
    "tm=[]\n",
    "Fwd = []\n",
    "Rev = []\n",
    "Len = []\n",
    "for i, gene in enumerate(genes2KO):\n",
    "    \n",
    "    #upstream PCR\n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.up')\n",
    "    Len.append(homology_length)\n",
    "    #downstream PCR\n",
    "    \n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.dw')\n",
    "    Len.append(homology_length)\n",
    "    \n",
    "df = np.stack([template, tube_label, tm , Len, Fwd, Rev ] ).T\n",
    "PCR_Table=pd.DataFrame(df , columns=[ 'Template', 'tube_label', 'tm','Product (bp)', 'Fwd','Rev'])\n",
    "\n",
    "PCR_Table.set_index('Template')\n",
    "\n",
    "PCR_Table.to_csv('/Users/paucac/downloads/sheet2del2.csv', sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Template</th>\n",
       "      <th>tube_label</th>\n",
       "      <th>tm</th>\n",
       "      <th>Product (bp)</th>\n",
       "      <th>Fwd</th>\n",
       "      <th>Rev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAX4</td>\n",
       "      <td>CAX4.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.001</td>\n",
       "      <td>BTX.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAX4</td>\n",
       "      <td>CAX4.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.003</td>\n",
       "      <td>BTX.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COX12</td>\n",
       "      <td>COX12.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.005</td>\n",
       "      <td>BTX.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COX12</td>\n",
       "      <td>COX12.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.007</td>\n",
       "      <td>BTX.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CUR1</td>\n",
       "      <td>CUR1.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.009</td>\n",
       "      <td>BTX.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CUR1</td>\n",
       "      <td>CUR1.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.011</td>\n",
       "      <td>BTX.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DLD2</td>\n",
       "      <td>DLD2.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.013</td>\n",
       "      <td>BTX.014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DLD2</td>\n",
       "      <td>DLD2.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.015</td>\n",
       "      <td>BTX.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>KGD2</td>\n",
       "      <td>KGD2.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.017</td>\n",
       "      <td>BTX.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>KGD2</td>\n",
       "      <td>KGD2.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.019</td>\n",
       "      <td>BTX.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MIP1</td>\n",
       "      <td>MIP1.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.021</td>\n",
       "      <td>BTX.022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MIP1</td>\n",
       "      <td>MIP1.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.023</td>\n",
       "      <td>BTX.024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NAM7</td>\n",
       "      <td>NAM7.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.025</td>\n",
       "      <td>BTX.026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NAM7</td>\n",
       "      <td>NAM7.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.027</td>\n",
       "      <td>BTX.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RCR2</td>\n",
       "      <td>RCR2.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.029</td>\n",
       "      <td>BTX.030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RCR2</td>\n",
       "      <td>RCR2.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.031</td>\n",
       "      <td>BTX.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SGM1</td>\n",
       "      <td>SGM1.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.033</td>\n",
       "      <td>BTX.034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SGM1</td>\n",
       "      <td>SGM1.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.035</td>\n",
       "      <td>BTX.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SHY1</td>\n",
       "      <td>SHY1.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.037</td>\n",
       "      <td>BTX.038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SHY1</td>\n",
       "      <td>SHY1.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.039</td>\n",
       "      <td>BTX.040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>VPS34</td>\n",
       "      <td>VPS34.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.041</td>\n",
       "      <td>BTX.042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>VPS34</td>\n",
       "      <td>VPS34.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.043</td>\n",
       "      <td>BTX.044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>YKU70</td>\n",
       "      <td>YKU70.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.045</td>\n",
       "      <td>BTX.046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>YKU70</td>\n",
       "      <td>YKU70.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.047</td>\n",
       "      <td>BTX.048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>YOR1</td>\n",
       "      <td>YOR1.up</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.049</td>\n",
       "      <td>BTX.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>YOR1</td>\n",
       "      <td>YOR1.dw</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>BTX.051</td>\n",
       "      <td>BTX.052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Template tube_label   tm Product (bp)      Fwd      Rev\n",
       "0      CAX4    CAX4.up  60C          400  BTX.001  BTX.002\n",
       "1      CAX4    CAX4.dw  60C          400  BTX.003  BTX.004\n",
       "2     COX12   COX12.up  60C          400  BTX.005  BTX.006\n",
       "3     COX12   COX12.dw  60C          400  BTX.007  BTX.008\n",
       "4      CUR1    CUR1.up  60C          400  BTX.009  BTX.010\n",
       "5      CUR1    CUR1.dw  60C          400  BTX.011  BTX.012\n",
       "6      DLD2    DLD2.up  60C          400  BTX.013  BTX.014\n",
       "7      DLD2    DLD2.dw  60C          400  BTX.015  BTX.016\n",
       "8      KGD2    KGD2.up  60C          400  BTX.017  BTX.018\n",
       "9      KGD2    KGD2.dw  60C          400  BTX.019  BTX.020\n",
       "10     MIP1    MIP1.up  60C          400  BTX.021  BTX.022\n",
       "11     MIP1    MIP1.dw  60C          400  BTX.023  BTX.024\n",
       "12     NAM7    NAM7.up  60C          400  BTX.025  BTX.026\n",
       "13     NAM7    NAM7.dw  60C          400  BTX.027  BTX.028\n",
       "14     RCR2    RCR2.up  60C          400  BTX.029  BTX.030\n",
       "15     RCR2    RCR2.dw  60C          400  BTX.031  BTX.032\n",
       "16     SGM1    SGM1.up  60C          400  BTX.033  BTX.034\n",
       "17     SGM1    SGM1.dw  60C          400  BTX.035  BTX.036\n",
       "18     SHY1    SHY1.up  60C          400  BTX.037  BTX.038\n",
       "19     SHY1    SHY1.dw  60C          400  BTX.039  BTX.040\n",
       "20    VPS34   VPS34.up  60C          400  BTX.041  BTX.042\n",
       "21    VPS34   VPS34.dw  60C          400  BTX.043  BTX.044\n",
       "22    YKU70   YKU70.up  60C          400  BTX.045  BTX.046\n",
       "23    YKU70   YKU70.dw  60C          400  BTX.047  BTX.048\n",
       "24     YOR1    YOR1.up  60C          400  BTX.049  BTX.050\n",
       "25     YOR1    YOR1.dw  60C          400  BTX.051  BTX.052"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('CTGCTCAAAAATATCCCATATATAC', 50.74756078704593)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " out_gap= 150\n",
    "dt.make_anneal(rec.seq[0:homology_length - out_gap], 'F' , 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#design the confirmation primers\n",
    "\n",
    "\n",
    "\n",
    "all_primers =[]\n",
    "all_temp = []\n",
    "position = []\n",
    "Gene= []\n",
    "NA =[]\n",
    "Primer_name=[]\n",
    "primer_n = 0\n",
    "\n",
    "for gene in genes2KO:\n",
    "    #open .gb file\n",
    "    start=0\n",
    "    stop=0\n",
    "    recs = [rec for rec in SeqIO.parse(path2parts + part_file[gene], \"genbank\")]\n",
    "    for rec in recs:\n",
    "        #loop through feature to features to find the location  of the CDS:\n",
    "        for feat in rec.features:\n",
    "            if gene + ' gene'in feat.qualifiers['label'] or gene in feat.qualifiers['label']:\n",
    "                location= feat.location\n",
    "                start = location.start.position\n",
    "                stop = location.end.position\n",
    "        #Design the confirmation out_gap:bp outside of homology region:\n",
    "        Up = dt.make_anneal(rec.seq[0:homology_length - out_gap], 'F' , 50)\n",
    "        Dw = dt.make_anneal(seq[0:homology_length + out_gap], 'R' , 50)\n",
    "        all_primers.append(Up[0])\n",
    "        all_temp.append(Up[1])\n",
    "        position.append('U.out')\n",
    "        Gene.append(gene)\n",
    "        \n",
    "        all_primers.append( Dw[0])\n",
    "        all_temp.append(Dw[1])\n",
    "        position.append('D.out')\n",
    "        Gene.append(gene)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CTGCATGTGACTGTATTAGTCA',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'AGCCATCATTTCGTTTAGTAGT',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'CCAATTGAAGAACTTTATGACAAC',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'AAAATACCAACCGTCCCT',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'TTTAGGATTTTGTAAAGAAAAAATGTCT',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'AAGACAAAGTTGTATCACTGTGA',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'CAAGGGTGGGTTATACCT',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'AAAAAGCTCAAGAAATCAACAAAAAG',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'AGGTCATCAGAACACAATGG',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'TTAACAGAGCTTGAAAATGAATCTAA',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'TTCTTAATCCTCTTCCCACC',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'GAAGTGCAAGGATTGATAATGTA',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'TCTGCTTCTTTGTTTAATGGTG',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'GTATTGCAAAAGGGTAATTAGC',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA',\n",
       " 'CTGCTCAAAAATATCCCATATATAC',\n",
       " 'ACTAACATTGTTGTATTGATGATAGA']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create primers for KOing TAL1 and ZWF1 in CCM strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['zwf1.gb', 'tal1.gb']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "path2parts = '/Users/paucac/Documents/PhD/CAB/CRISPA/CCM/DNA_parts/KO_confirmations/'\n",
    "os.listdir(path2parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## dict storing file names of parts:\n",
    "genes2KO=['zwf1','tal1']\n",
    "genes2KO= [g.upper() for g in genes2KO]\n",
    "part_file={}\n",
    "for gene in genes2KO:\n",
    "    part_file[gene]= gene+\".gb\"\n",
    "\n",
    "part_seq={}\n",
    "part_seq['KanMX']='taggtctagagatctgtttagcttgcctcgtccccgccgggtcacccggccagcgacatggaggcccagaataccctccttgacagtcttgacgtgcgcagctcaggggcatgatgtgactgtcgcccgtacatttagcccatacatccccatgtataatcatttgcatccatacattttgatggccgcacggcgcgaagcaaaaattacggctcctcgctgcagacctgcgagcagggaaacgctcccctcacagacgcgttgaattgtccccacgccgcgcccctgtagagaaatataaaaggttaggatttgccactgaggttcttctttcatatacttccttttaaaatcttgctaggatacagttctcacatcacatccgaacataaacaaccatgggtaaggaaaagactcacgtttcgaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccggcaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataagcttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaatcagtactgacaataaaaagattcttgttttcaagaacttgtcatttgtatagtttttttatattgtagttgttctattttaatcaaatgttagcgtgatttatattttttttcgcctcgacatcatctgcccagatgcgaagttaagtgcgcagaaagtaatatcatgcgtcaatcgtatgtgaatgctggtcgctatactgctgtcgattcgatactaacgccgccatccagtgtcgaaaacgagctctcgagaacccttaat'\n",
    "\n",
    "#Defines the Overhangs with KanMX  \n",
    "KanMX_overhang_Up = dt.reverse_complement(part_seq['KanMX'][0:40]) #we copy the reverse complement for the upstream up\n",
    "KanMX_overhang_Dw = part_seq['KanMX'][-40:]\n",
    "\n",
    "\n",
    "#extract the sequence [upper homology:CDS:lower homology]\n",
    "\n",
    "for gene in genes2KO:\n",
    "    #open .gb file\n",
    "    start=0\n",
    "    stop=0\n",
    "    recs = [rec for rec in SeqIO.parse(path2parts + part_file[gene], \"genbank\")]\n",
    "    for rec in recs:\n",
    "        #loop through feature to features to find the location  of the CDS:\n",
    "        for feat in rec.features:\n",
    "            if gene + ' gene'in feat.qualifiers['label']:\n",
    "                location= feat.location\n",
    "                start = location.start.position\n",
    "                stop = location.end.position\n",
    "        #stores seq in dict\n",
    "        part_seq[gene] = str(rec.seq[start-homology_length:stop+homology_length])\n",
    "\n",
    "\n",
    "        \n",
    "#design the KanMX primers\n",
    "Kan_up= dt.make_anneal(part_seq['KanMX'].upper(), 'F' , 50)\n",
    "Kan_dw= dt.make_anneal(part_seq['KanMX'], 'R' , 50)\n",
    "\n",
    "#designs the primers for KO:\n",
    "importlib.reload(dt)\n",
    "UF_list=[]\n",
    "UR_list=[]\n",
    "DF_list=[]\n",
    "DR_list=[]\n",
    "for gene in genes2KO:\n",
    "    #design the upstream F and R primers\n",
    "    seq= part_seq[gene]\n",
    "    UF = dt.make_anneal(seq, 'F' , 50)\n",
    "    UR = dt.make_anneal(seq[0:homology_length], 'R' , 50)\n",
    "    #design the downstream F and R primers:\n",
    "    DF = dt.make_anneal(seq[-homology_length:], 'F' , 50)\n",
    "    DR = dt.make_anneal(seq, 'R' , 50)\n",
    "\n",
    "    #copy the KanMX 30bp overhang to the UR and DF:\n",
    "    UR =  (( KanMX_overhang_Up).lower() + UR[0], UR[1])\n",
    "\n",
    "    DF =  ((KanMX_overhang_Dw).lower() + DF[0], DF[1])\n",
    "    \n",
    "    UF_list.append(UF)\n",
    "    UR_list.append(UR)\n",
    "    DF_list.append(DF)\n",
    "    DR_list.append(DR)\n",
    "    \n",
    "#sort the ouptut into a nice data frame\n",
    "\n",
    "all_primers =[]\n",
    "all_temp =[]\n",
    "position = []\n",
    "Gene= []\n",
    "NA =[]\n",
    "Primer_name=[]\n",
    "\n",
    "primer_n = 77 #to match your current primer indexing\n",
    "for i, gene in enumerate(genes2KO):\n",
    "\n",
    "        \n",
    "    all_primers.append(UF_list[i][0])\n",
    "    all_temp.append(str(round(UF_list[i][1] ,1) +6 ) + ' phus')      #we add 6C for NEB phusion\n",
    "    position.append('UF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('CCM.'+str(i+1+3*i +primer_n ).zfill(3))\n",
    "\n",
    "    \n",
    "    all_primers.append(UR_list[i][0])\n",
    "    all_temp.append(str(round(UR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('UR')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('CCM.'+str(i+2+3*i +primer_n).zfill(3))\n",
    "    \n",
    "    all_primers.append(DF_list[i][0])\n",
    "    all_temp.append(str(round(DF_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DF')\n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('CCM.'+str(i+3+3*i +primer_n).zfill(3))\n",
    "    \n",
    "    all_primers.append(DR_list[i][0])\n",
    "    all_temp.append(str(round(DR_list[i][1],1) + 6)+ ' phus')   \n",
    "    position.append('DR')               \n",
    "    Gene.append(gene)\n",
    "    NA.append('na')\n",
    "    Primer_name.append('CCM.'+str(i+4+3*i +primer_n).zfill(3))\n",
    " \n",
    "#add the KanMX primers\n",
    "all_primers.append(Kan_up[0])\n",
    "all_temp.append(str(round(Kan_up[1],1) + 6)+ ' phus')   \n",
    "position.append('F')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "all_primers.append(Kan_dw[0])\n",
    "all_temp.append(str(round(Kan_dw[1],1) + 6)+ ' phus')   \n",
    "position.append('R')\n",
    "Gene.append('KanMX (pCfB 2312)')\n",
    "NA.append('na')\n",
    "Primer_name.append('BTX.'+'XXX')\n",
    "\n",
    "\n",
    "#some primers are over IDT 60bp limit we shorten there homology region to KanMX:\n",
    "for i, primer in enumerate(all_primers):\n",
    "    if len(primer) >60:\n",
    "        all_primers[i] = primer[-60:]\n",
    "        if  (sum([bp.islower() for bp in all_primers[i]]) < 30):\n",
    "            print('WARNING your homology overhang was cut to <30bp')                    \n",
    "                            \n",
    "df = np.stack([Primer_name, all_primers,position, all_temp,NA, Gene, NA, NA] ).T\n",
    "Primer_Table=pd.DataFrame(df , columns=['Primer Name', 'Seq','Dir', 'Tm (c)','User', 'Template', 'Template_Len','benchling'])\n",
    "\n",
    "Primer_Table.set_index('Primer Name')\n",
    "\n",
    "Primer_Table.to_csv('/Users/paucac/downloads/sheet2del.csv', sep=',')        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tube_label</th>\n",
       "      <th>Template</th>\n",
       "      <th>tm</th>\n",
       "      <th>Product (bp)</th>\n",
       "      <th>Fwd</th>\n",
       "      <th>Rev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZWF1.up</td>\n",
       "      <td>ZWF1</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>CCM.078</td>\n",
       "      <td>CCM.079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZWF1.dw</td>\n",
       "      <td>ZWF1</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>CCM.080</td>\n",
       "      <td>CCM.081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TAL1.up</td>\n",
       "      <td>TAL1</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>CCM.082</td>\n",
       "      <td>CCM.083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TAL1.dw</td>\n",
       "      <td>TAL1</td>\n",
       "      <td>60C</td>\n",
       "      <td>400</td>\n",
       "      <td>CCM.084</td>\n",
       "      <td>CCM.085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Tube_label Template   tm Product (bp)      Fwd      Rev\n",
       "0    ZWF1.up     ZWF1  60C          400  CCM.078  CCM.079\n",
       "1    ZWF1.dw     ZWF1  60C          400  CCM.080  CCM.081\n",
       "2    TAL1.up     TAL1  60C          400  CCM.082  CCM.083\n",
       "3    TAL1.dw     TAL1  60C          400  CCM.084  CCM.085"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create a PCR table\n",
    "tube_label = []\n",
    "template = []\n",
    "tm=[]\n",
    "Fwd = []\n",
    "Rev = []\n",
    "Len = []\n",
    "for i, gene in enumerate(genes2KO):\n",
    "    \n",
    "    #upstream PCR\n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'UR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.up')\n",
    "    Len.append(homology_length)\n",
    "    #downstream PCR\n",
    "    \n",
    "    template.append(gene)\n",
    "    tm.append('60C')\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DF')\n",
    "    Fwd.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    i = (Primer_Table['Template']== gene) & (Primer_Table['Dir']== 'DR')\n",
    "    Rev.append(Primer_Table['Primer Name'][i].values[0])\n",
    "    tube_label.append(gene +'.dw')\n",
    "    Len.append(homology_length)\n",
    "    \n",
    "df = np.stack([ tube_label, template,tm , Len, Fwd, Rev ] ).T\n",
    "PCR_Table=pd.DataFrame(df , columns=[  'Tube_label','Template', 'tm','Product (bp)', 'Fwd','Rev'])\n",
    "\n",
    "PCR_Table.set_index('Template')\n",
    "\n",
    "PCR_Table.to_csv('/Users/paucac/downloads/sheet2del2.csv', sep=',')\n",
    "PCR_Table"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DNAEnv",
   "language": "python",
   "name": "dnaenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
